File: 2021/03/main.go
package main

import (
	"fmt"
	"os"
	"bufio"
	"strings"
	"strconv"
)

func get_readings(file *os.File) []string {
	var readings []string

	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		line := scanner.Text()
		readings = append(readings, line)
	}
	if err := scanner.Err(); err != nil {
		fmt.Fprintln(os.Stderr, "reading standard input:", err)
	}
	
	return readings
}

func get_most_common_bit(bits [][]string, position int) map[string]int {
	zero_count := 0
	one_count := 0

	for _, bit := range(bits) {
		if bit[position] == "0" {
			zero_count ++
		}
		if bit[position] == "1" {
			one_count ++
		}
	}

	return map[string]int{"0": zero_count, "1": one_count}
}

func part_one(readings []string) {
	var bits [][]string

	// Get list of lists of bits in each number
	for _, line := range(readings) {
		bit := strings.Split(line, "")
		bits = append(bits, bit)
	}

	// Get list of maps of how much each bit appeared
	var most_common_bits []map[string]int
	for i := 0; i < len(bits[0]); i++ {
		individual_most_common_bit := get_most_common_bit(bits, i)
		most_common_bits = append(most_common_bits, individual_most_common_bit)
	}

	// Get most and least common bit in each position
	var most_common_bit string
	var least_common_bit string

	for _, bit_mode := range(most_common_bits) {
		if bit_mode["0"] > bit_mode["1"] {
			most_common_bit += "0"
			least_common_bit += "1"
		}
		if bit_mode["0"] < bit_mode["1"]{
			most_common_bit += "1"
			least_common_bit += "0"
		}
	}

	fmt.Println("Most common bit: ", most_common_bit)
	gamma_rate, _ := strconv.ParseInt(most_common_bit, 2, 64)
	fmt.Println("Gamma rate: ", gamma_rate)

	fmt.Println("Least common bit: ", least_common_bit)
	epsilon_rate, _ := strconv.ParseInt(least_common_bit, 2, 64)
	fmt.Println("Epsilon rate: ", epsilon_rate)

	power_consumption := gamma_rate * epsilon_rate
	fmt.Println("Power consumption: ", power_consumption)
}


func part_two(readings []string) {
	bit_mode := make(map[string]int)
	for _, line := range(readings) {
		bit := string(line[0])
		bit_mode[bit] ++
		
	}

	if bit_mode["0"] > bit_mode["1"] {
		most_common_bit += "0"
		least_common_bit += "1"
	}
	if bit_mode["0"] < bit_mode["1"] || bit_mode["0"] == bit_mode["1"] { // TODO change to else?
		most_common_bit += "1"
		least_common_bit += "0"
	}
}


func main() {
	readings := get_readings(os.Stdin)

	// part_one(readings)
	part_two(readings)

	// Part 2
	// oxygen_generator_rating, _ := strconv.ParseInt(most_common_part_two_bit, 2, 64)
	// fmt.Println("Oxygen generator rating: ", oxygen_generator_rating)

	// co2_scrubber_rating, _ := strconv.ParseInt(least_common_part_two_bit, 2, 64)
	// fmt.Println("CO2 scrubber rating: ", co2_scrubber_rating)
}

File: 2021/03/python/main.py
import sys
from collections import Counter


def part_one(report):
    most_common_bits, least_common_bits = "", ""
    for i in range(0, len(report[0])):
        bit_appearances = Counter([reading[i] for reading in report]).most_common()
        most_common_bits += bit_appearances[0][0]
        least_common_bits += bit_appearances[-1][0]

    gamma_rate = int(most_common_bits, 2)
    epsilon_rate = int(least_common_bits, 2)
    power_consumption = gamma_rate * epsilon_rate

    return power_consumption


def part_two(report):
    most_common_report_copy = report
    least_common_report_copy = report

    for position in range(0, len(most_common_report_copy[0])):
        char_frequency = Counter([reading[position] for reading in most_common_report_copy]).most_common()
        if char_frequency[0][1] == char_frequency[1][1]:
            most_common_char = "1"
        else:
            most_common_char = char_frequency[0][0]
        most_common_report_copy = [i for i in most_common_report_copy if i[position] == most_common_char]
        if len(most_common_report_copy) == 1:
            break

    for position in range(0, len(least_common_report_copy[0])):
        char_frequency = Counter([reading[position] for reading in least_common_report_copy]).most_common()
        if char_frequency[0][1] == char_frequency[1][1]:
            least_common_char = "0"
        else:
            least_common_char = char_frequency[1][0]
        least_common_report_copy = [i for i in least_common_report_copy if i[position] == least_common_char]
        if len(least_common_report_copy) == 1:
            break

    oxygen_generator_rating_binary = most_common_report_copy[0]
    oxygen_generator_rating_decimal = int(oxygen_generator_rating_binary, 2)

    co2_scrubber_rating_binary = least_common_report_copy[0]
    co2_scrubber_rating_decimal = int(co2_scrubber_rating_binary, 2)

    life_support_rating = oxygen_generator_rating_decimal * co2_scrubber_rating_decimal

    return life_support_rating


def main():
    path_to_file = sys.argv[1]
    with open(path_to_file) as f:
        report = [line.strip() for line in f.readlines()]
    print(f"Part One: {part_one(report)}")
    print(f"Part Two: {part_two(report)}")


if __name__ == "__main__":
    main()
